src/resource.rs

```diff
@@ -400,34 +400,28 @@ where
             *rdef.name_mut() = name.clone();
         }
 
-        config.register_service(rdef, guards, self, None)
-    }
-}
-
-impl<T> IntoServiceFactory<T, ServiceRequest> for Resource<T>
-where
-    T: ServiceFactory<
-        ServiceRequest,
-        Config = (),
-        Response = ServiceResponse,
-        Error = Error,
-        InitError = (),
-    >,
-{
-    fn into_factory(self) -> T {
         *self.factory_ref.borrow_mut() = Some(ResourceFactory {
             routes: self.routes,
-            app_data: self.app_data.map(Rc::new),
             default: self.default,
         });
 
-        self.endpoint
+        let resource_data = self.app_data.map(Rc::new);
+
+        // wraps endpoint service (including middleware) call and injects app data for this scope
+        let endpoint = apply_fn_factory(self.endpoint, move |mut req: ServiceRequest, srv| {
```

Comment:
Is the type annotation (on `req`) needed?

Reply:
yep

```
454 |         let endpoint = apply_fn_factory(self.endpoint, move |mut req, srv| {
    |                                                              ^^^^^^^ consider giving this closure parameter a type
```